### PR TITLE
udevadm-verify: add support for directory arguments

### PR DESCRIPTION
**Pull request by ldv-alt**
When an argument specified to `udevadm verify` is a directory, verify all *.rules files in that directory.

Suggested-by: Zbigniew Jędrzejewski-Szmek <zbyszek@in.waw.pl>

### CODE REVIEW

File: src/udev/udevadm-verify.c [hunk 1]
```diff
@@ -119,20 +119,45 @@ static int verify_rules_file(UdevRules *rules, const char *fname) {
...
 
         STRV_FOREACH(fp, files) {
-                r = verify_rules_file(rules, *fp);
-                if (r < 0) {
-                        fail_count++;
-                        if (rv >= 0)
-                                rv = r;
-                } else
-                        success_count++;
+                if (endswith(*fp, "/"))
```
Comment (keszybz) [inline src/udev/udevadm-verify.c]: I don't think this is very user friendly. The decision should be based on whether the path is really a directory, and not based on how the user specified it. (I know I wrote something like this in telegram, but that was just a telegram shorthand… We don't have logic like this in other places in the codebase.)

Comment (ldv-alt) [reply src/udev/udevadm-verify.c]: OK, let's just descent into directories.

File: src/udev/udevadm-verify.c [hunk 2]
```diff
@@ -119,20 +119,47 @@ static int verify_rules_file(UdevRules *rules, const char *fname) {
...
-                r = verify_rules_file(rules, *fp);
-                if (r < 0) {
-                        fail_count++;
-                        if (rv >= 0)
-                                rv = r;
-                } else
-                        success_count++;
+                struct stat st;
+
+                if (stat(*fp, &st) == 0 && S_ISDIR(st.st_mode))
```
Comment (yuwata) [inline src/udev/udevadm-verify.c]: We have path_is_dir() in stat-util.h (the name may be slightly different, as I am now writing this from phone.)

**Review by yuwata — COMMENTED** [review]
Please add assertions for arguments in each function, even that's trivial.

File: src/udev/udevadm-verify.c (hunk 2 above)
Comment (ldv-alt) [reply src/udev/udevadm-verify.c]: Its name is `is_dir`.

File: src/udev/udevadm-verify.c (hunk 1 above)
Comment (ldv-alt) [reply src/udev/udevadm-verify.c]: But in that case extra care should be taken to avoid directory loops.

**Comment by keszybz** [comment]
```console
$ build/udevadm verify /no/such/directory
Failed to open /no/such/directory, ignoring: No such file or directory
Failed to parse rules file /no/such/directory: No such file or directory

1 udev rules files have been checked.
  Success: 0
  Fail:    1
```
I think that if there's an explicit argument, it should just fail. But it certainly shouldn't log twice (once at warning level, once at error level).

File: src/udev/udevadm-verify.c [hunk 3]
```diff
@@ -119,20 +120,58 @@ static int verify_rules_file(UdevRules *rules, const char *fname) {
...
+                if (r < 0 && rv >= 0)
+                        rv = r;
         }
 
+        return rv;
+}
+
+static int verify_rules(UdevRules *rules, char **files) {
+        size_t fail_count = 0, success_count = 0;
+        int rv;
```
Comment (keszybz) [inline src/udev/udevadm-verify.c]: Please just call it `r` following the usual style.

**Comment by ldv-alt** [comment]
> ```
> $ build/udevadm verify /no/such/directory
> Failed to open /no/such/directory, ignoring: No such file or directory
> Failed to parse rules file /no/such/directory: No such file or directory
> 
> 1 udev rules files have been checked.
>   Success: 0
>   Fail:    1
> ```
> 
> I think that if there's an explicit argument, it should just fail. But it certainly shouldn't log twice (once at warning level, once at error level).

I agree, but that is not related to this PR.

File: src/udev/udev-rules.c [hunk 4]
```diff
@@ -1454,7 +1454,10 @@ int udev_rules_parse_file(UdevRules *rules, const char *filename, bool extra_che
 
         f = fopen(filename, "re");
         if (!f) {
-                if (!extra_checks && errno == ENOENT)
+                if (extra_checks)
+                        return -ERRNO_VALUE(errno);
```
Comment (yuwata) [inline src/udev/udev-rules.c]: Why `ERRNO_VALUE()` is used here? Please simply use `return -errno;` as usual.

**Review by yuwata — APPROVED** [review]
LGTM, except for one in below.

**Pull request merged**